File: gui/file_explorer.py
from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, QLineEdit, QHBoxLayout, QTreeWidgetItem, QHeaderView,
    QLabel, QMenuBar, QAction, QTreeWidget
)
from PyQt5.QtGui import QPixmap
from PyQt5.QtCore import Qt, QTimer, QSize
from settings import Settings
from actions import Actions
import fileops, os

class FileExplorer(QWidget):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("QtFiles")
        self.setFixedSize(800, 400)
        self.settings = fileops.Settings()
        self.init_ui()
        self.load_files(self.current_path)
        self.actions = Actions(self)

    def init_ui(self):
        layout = QVBoxLayout(self)
        
        self.menubar = QMenuBar(self)
        settings_menu = self.menubar.addMenu("File")

        open_settings_action = QAction("Settings", self)
        open_settings_action.triggered.connect(self.open_settings)
        settings_menu.addAction(open_settings_action)
        
        layout.setMenuBar(self.menubar)

        self.icon_cache = {
            "folder_filled": QPixmap("gui/assets/folder_filled.png"),
            "folder_empty": QPixmap("gui/assets/folder_empty.png"),
            "file": QPixmap("gui/assets/file.png")
        }

        self.current_path = fileops.get_home_directory()

        self.directory_display = QLineEdit()
        self.directory_display.setText(self.current_path)
        self.directory_display.returnPressed.connect(self.on_directory_entered)

        self.search_bar = QLineEdit()
        self.search_bar.setFixedWidth(200)
        self.search_bar.setPlaceholderText("Search")
        self.search_bar.returnPressed.connect(self.on_search_enter)

        self.top_bar = QHBoxLayout()
        self.top_bar.addWidget(self.directory_display)
        self.top_bar.addWidget(self.search_bar)

        layout.addLayout(self.top_bar)

        self.tree_widget = QTreeWidget()
        self.tree_widget.setRootIsDecorated(False) # For now
        self.tree_widget.setColumnCount(2)
        self.tree_widget.setHeaderLabels(["Name", "Size"])
        self.tree_widget.setStyleSheet(f"QTreeWidget::item:selected {{ background-color: {self.settings.get_setting('selected_bg_color')}; }}")

        header = self.tree_widget.header()
        header.setSectionResizeMode(0, QHeaderView.Stretch)
        header.setSectionResizeMode(1, QHeaderView.ResizeToContents) 

        layout.addWidget(self.tree_widget)
        
        self.files_to_render = []
        self.tree_widget.itemDoubleClicked.connect(self.on_item_double_clicked)

    def load_files(self, fp):
        if not isinstance(fp, str):
            fp = fileops.get_home_directory()

        self.current_path = fp
        self.directory_display.setText(self.current_path)
        self.tree_widget.clear()

        files = fileops.list_files(fp, False)
        self.files_to_render = files
        self.render_next_batch()

    def render_next_batch(self):
        BATCH_SIZE = 100
        count = 0

        while self.files_to_render and count < BATCH_SIZE:
            file = self.files_to_render.pop(0)

            file_widget = QWidget()
            file_layout = QHBoxLayout(file_widget)
            file_layout.setContentsMargins(4, 2, 4, 2)
            file_layout.setSpacing(6)

            icon_size_int = int(self.settings.get_setting("icon_size"))
            item_height = int(self.settings.get_setting("item_height"))
            
            if file.is_directory:
                if fileops.check_for_dir_contents(os.path.join(self.current_path, file.filename)):
                    icon_to_load = "folder_filled"
                else:
                    icon_to_load = "folder_empty"
            else:
                icon_to_load = "file"

            icon_label = QLabel()
            icon_label.setPixmap(self.icon_cache[icon_to_load])
            icon_label.setFixedSize(icon_size_int, icon_size_int)
            icon_label.setScaledContents(True)
            
            font_color = self.settings.get_setting('font_color')
            font_size = self.settings.get_setting('font_size')

            filename_label = QLabel(file.filename)
            filename_label.setStyleSheet(f"font-size: {font_size}px; color: {font_color};")

            filesize_label = QLabel(file.filesize)
            filesize_label.setStyleSheet(f"font-size: {font_size}px; color: {font_color};")

            file_layout.addWidget(icon_label)
            file_layout.addWidget(filename_label)
            file_layout.addStretch()

            item = QTreeWidgetItem(["", ""])
            item.setSizeHint(0, QSize(0, item_height))

            item.setData(0, Qt.UserRole, os.path.join(self.current_path, file.filename))
            item.setData(0, Qt.UserRole + 1, bool(file.is_directory))

            self.tree_widget.addTopLevelItem(item)
            self.tree_widget.setItemWidget(item, 0, file_widget)
            self.tree_widget.setItemWidget(item, 1, filesize_label)

            count += 1

        if self.files_to_render:
            QTimer.singleShot(10, self.render_next_batch)

    def on_item_double_clicked(self, item, column):
        filepath = item.data(0, Qt.UserRole)
        is_directory = item.data(0, Qt.UserRole + 1)

        if is_directory:
            self.load_files(filepath)

    def on_directory_entered(self):
        self.load_files(self.directory_display.text())

    def on_search_enter(self):
        search_text = self.search_bar.text()
        fileops_result = fileops.search_in_dir(self.current_path, search_text)

        self.tree_widget.clear()
        self.files_to_render = fileops_result
        self.render_next_batch()
        
    def open_settings(self):
        Settings()
        self.load_files(self.current_path)
        
    def refresh(self):
        self.load_files(self.current_path)

File: gui/file_explorer.cpp
#include "file_explorer.h"

#include <iostream>


FileExplorer::FileExplorer(): QWidget()
{
  setWindowTitle("QtFiles");
  setFixedSize(800, 400); //Maybe dont do that.
  initUI();

  this->modal_builder = ModalBuilder();

  // Open Explorer in Users home directory
  current_path = getHomeDirectory();
  loadFiles(QString::fromStdString(current_path));
}

void FileExplorer::initUI()
{
  this->setContextMenuPolicy(Qt::CustomContextMenu);
  connect(this, &QWidget::customContextMenuRequested, this, &FileExplorer::showContextMenu);

  main_layout = new QVBoxLayout(this);
  menu_bar = new QMenuBar(this);
  QMenu *settings_menu = menu_bar->addMenu("File");

  QAction *open_settings_action = new QAction("Settings", this);
  connect(open_settings_action, &QAction::triggered, this, &FileExplorer::openSettings);
  settings_menu->addAction(open_settings_action);

  main_layout->setMenuBar(menu_bar);

  icon_cache["folder_filled"] = QPixmap("gui/assets/folder_filled.png");
  icon_cache["folder_empty"] = QPixmap("gui/assets/folder_empty.png");
  icon_cache["file"] = QPixmap("gui/assets/file.png");

  directory_display = new QLineEdit(this);
  connect(directory_display, &QLineEdit::returnPressed, this, &FileExplorer::onDirectoryEntered);

  search_bar = new QLineEdit(this);
  search_bar->setFixedWidth(200);
  search_bar->setPlaceholderText("Search for file");
  connect(search_bar, &QLineEdit::returnPressed, this, &FileExplorer::onSearchEntered);

  top_bar = new QHBoxLayout();
  top_bar->addWidget(directory_display);
  top_bar->addWidget(search_bar);
  main_layout->addLayout(top_bar);

  // Tree widget
  tree_widget = new QTreeWidget(this);
  tree_widget->setRootIsDecorated(false);
  tree_widget->setColumnCount(2);
  tree_widget->setHeaderLabels({"Name", "Size"});

  QHeaderView *header = tree_widget->header();
  header->setSectionResizeMode(0, QHeaderView::Stretch);
  header->setSectionResizeMode(1, QHeaderView::ResizeToContents);

  connect(tree_widget, &QTreeWidget::itemDoubleClicked,
          this, &FileExplorer::onItemDoubleClicked);

  main_layout->addWidget(tree_widget);
}

void FileExplorer::loadFiles(const QString &path)
{
  current_path = path.toStdString();
  directory_display->setText(path);
  
  auto entries = listFiles(current_path);

  createList(entries);
}


void FileExplorer::createList(auto entries, ListMode mode) {
  tree_widget->clear();

  for (const auto &file : entries) {
    QString icon_key = file.type == FileType::Directory ? "folder_filled" : "file";

    QWidget *file_widget = new QWidget();
    QHBoxLayout *file_layout = new QHBoxLayout(file_widget);
    file_layout->setContentsMargins(4, 2, 4, 2);
    file_layout->setSpacing(6);

    QLabel *icon_label = new QLabel();
    icon_label->setPixmap(icon_cache[icon_key]);
    icon_label->setFixedSize(16, 16);
    icon_label->setScaledContents(true);

    std::string filename = (mode == ListMode::Search) ? file.path : file.name;

    QLabel *filename_label = new QLabel(QString::fromStdString(filename));
    filename_label->setObjectName("filename_label");
    file_layout->addWidget(icon_label);
    file_layout->addWidget(filename_label);
    file_layout->addStretch();

    QLabel *filesize_label = new QLabel(formatByte(file.size));

    QTreeWidgetItem *item = new QTreeWidgetItem({"", ""});
    item->setData(0, Qt::UserRole, QString::fromStdString(filename));
    item->setData(0, Qt::UserRole + 1, file.type == FileType::Directory);

    tree_widget->addTopLevelItem(item);
    tree_widget->setItemWidget(item, 0, file_widget);
    tree_widget->setItemWidget(item, 1, filesize_label);
  }
}

void FileExplorer::keyPressEvent(QKeyEvent *event) {
  QTreeWidgetItem *item = tree_widget->currentItem();

  switch(event->key()) 
  {
    case Qt::Key_F2: {
      if (item) {
        handleRename(item);
      }  
    }
    case Qt::Key_Delete: {
      if (item) {
        handleDelete(item);
      }
    }
    default:
      QWidget::keyPressEvent(event);
  }
}

void FileExplorer::showContextMenu(const QPoint &pos) 
{
  QAction *create_file_action = new QAction("File", this);
  QAction *create_directory_action = new QAction("Directory", this);

  // Using lambda, since we need the argument what we want to create.
  connect(create_file_action, &QAction::triggered, this,
    [this]() { handleEntityCreate(FileType::File); });

  connect(create_directory_action, &QAction::triggered, this,
    [this]() { handleEntityCreate(FileType::Directory); });
  QMenu context_menu(tr("Context menu"), this);

  QMenu new_entity_menu("New", this);
  new_entity_menu.addAction(create_file_action);
  new_entity_menu.addAction(create_directory_action);

  context_menu.addMenu(&new_entity_menu);

  context_menu.exec(mapToGlobal(pos));
}

void FileExplorer::handleEntityCreate(FileType type) {
  QString label_text = QString("Create new ") + QString(type == FileType::File ? "File" : "Directory");
  std::string name = this->modal_builder.showInputModal(this, label_text, "Name").toStdString();

  if (name != "") {
    std::filesystem::path path(current_path + PATH_SEPARATOR + name);
    bool created_entity = createFile(path, type);

    if (!created_entity) {
      QString error_message = QString("Failed to create ") + QString(type == FileType::File ? "file." : "directory.");
      this->modal_builder.showErrorModal(error_message);

      return;
    }

    loadFiles(QString::fromStdString(current_path));
  }
}

void FileExplorer::handleRename(QTreeWidgetItem *item) {
  QWidget *file_widget = tree_widget->itemWidget(item, 0);
  if (!file_widget) return;

  QHBoxLayout *layout = qobject_cast<QHBoxLayout*>(file_widget->layout());
  if (!layout) return;

  QLabel *filename_label = file_widget->findChild<QLabel *>("filename_label");
  if (!filename_label) return;

  QString filename = filename_label->text();

  QLineEdit *edit = new QLineEdit(filename, file_widget);
  edit->selectAll();
  layout->replaceWidget(filename_label, edit);
  filename_label->hide();

  edit->setFocus();

  connect(edit, &QLineEdit::editingFinished, this, [this, filename, edit, filename_label, item]() {
    QString new_name = edit->text().trimmed();

    if (!new_name.isEmpty() && new_name != filename_label->text()) {
      std::string filepath = current_path + PATH_SEPARATOR + filename.toStdString();

      std::string new_path = current_path + PATH_SEPARATOR + new_name.toStdString();

      renameFile(filepath, new_path);
    }

    // Reload current directory after changing filenames
    loadFiles(QString::fromStdString(current_path));
  });
}

void FileExplorer::handleDelete(QTreeWidgetItem *item) {
  QWidget *file_widget = tree_widget->itemWidget(item, 0);
  if (!file_widget) return;

  QHBoxLayout *layout = qobject_cast<QHBoxLayout*>(file_widget->layout());
  if (!layout) return;

  QLabel *filename_label = file_widget->findChild<QLabel *>("filename_label");
  if (!filename_label) return;

  QString filename = filename_label->text();
  std::string filepath = current_path + PATH_SEPARATOR + filename.toStdString();
  bool deleted = deleteFile(filepath);

  if (!deleted) {
    this->modal_builder.showErrorModal("Failed to delete the file or directory.");
    return;
  }

  loadFiles(QString::fromStdString(current_path));
}

void FileExplorer::onDirectoryEntered()
{
  std::string dir_path = directory_display->text().toStdString();
  std::filesystem::path path(dir_path);

  if (!std::filesystem::exists(path) || !std::filesystem::is_directory(path)) {
    this->modal_builder.showErrorModal("The specified path does not exist.");
    directory_display->setText(QString::fromStdString(current_path));
    return;
  }

  loadFiles(QString::fromStdString(dir_path));
}

void FileExplorer::onItemDoubleClicked(QTreeWidgetItem *item, int column)
{
  QString filepath = item->data(0, Qt::UserRole).toString();
  bool is_directory = item->data(0, Qt::UserRole + 1).toBool();

  if (is_directory) {
    std::string filepath_string = filepath.toStdString();
    if (filepath_string.find(current_path) == std::string::npos) {
      filepath_string = current_path + PATH_SEPARATOR + filepath_string;
    }

    loadFiles(QString::fromStdString(filepath_string));
  }
}

void FileExplorer::onSearchEntered()
{
  std::string query = search_bar->text().toStdString();
  auto search_result = searchDirectory(current_path, query);

  createList(search_result, ListMode::Search);
}

/**
 * We'll ignore settings for now.
 */
void FileExplorer::openSettings()
{
  qDebug() << "Open settings (Not implemented yet)";
  //open settings
}
